src/librustc_trans/mir/rvalue.rs

```diff
@@ -272,6 +272,29 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
                         let r_t_out = CastTy::from_ty(cast.ty).expect("bad output type for cast");
                         let ll_t_in = operand.layout.immediate_llvm_type(bx.cx);
                         let ll_t_out = cast.immediate_llvm_type(bx.cx);
+                        match operand.layout.variants {
+                            layout::Variants::Single { index } => {
+                                let variant = operand.layout.for_variant(bx.cx, index);
+                                if variant.abi == layout::Abi::Uninhabited {
+                                    return (bx, OperandRef {
+                                        val: OperandValue::Immediate(C_undef(ll_t_out)),
+                                        layout: cast,
+                                    });
+                                }
+                                if let Some(def) = operand.layout.ty.ty_adt_def() {
+                                    let discr_val = def
+                                        .discriminant_for_variant(bx.cx.tcx, index)
+                                        .val as u64;
+                                    let discr = C_uint(ll_t_out, discr_val);
+                                    return (bx, OperandRef {
+                                        val: OperandValue::Immediate(discr),
+                                        layout: cast,
+                                    });
```

Comment:
Similar to miri, you should preferably use `trans_get_discr`.

Reply:
I don't have a `Place` there, so I can't use `trans_get_discr`.